ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,409 @@
+                    nd = len(tp)
+                    rc = tp / non_ignored
+                    pr = tp / (fp + tp + torch.finfo(torch.float64).eps)
+                    q: Any = torch.zeros((num_rec_thr,))
+
+                    pr = pr.tolist()
+                    q = q.tolist()
+
+                    for i in range(nd - 1, 0, -1):
+                        if pr[i] > pr[i - 1]:
+                            pr[i - 1] = pr[i]
+
+                    inds = torch.searchsorted(rc, self.rec_thresholds, right=False)
+                    try:
+                        for ri, pi in enumerate(inds):
+                            q[ri] = pr[pi]
+                    except:
+                        pass
+                    precision[t, :, c, a] = torch.tensor(q)
+        # return precision, recall
+        self.precision = precision
+
+    def _summarize(self, iou_thr: Optional[float] = None, area_rng: str = "all") -> float:
+        aind = [i for i, a_rng in enumerate(self.object_area_ranges) if a_rng == area_rng]
+        # Calculate Average Precision
+        s = self.precision
+        if iou_thr is not None:
+            t = (self.iou_thresholds == iou_thr).int().nonzero(as_tuple=True)[0]
+            s = s[t]
+        s = s[:, :, :, aind]
+        # Take mean to calculate mAP
+        if len(s[s > -1]) == 0:
+            mean_s = torch.tensor(-1)
+        else:
+            mean_s = torch.mean(s[s > -1])
+
+        return mean_s.item()
+
+    @reinit__is_reduced
+    def _gather_all(self) -> None:
+        import torch.distributed as dist
+
+        def is_dist_avail_and_initialized() -> bool:
+            if not dist.is_available():
+                return False
+            if not dist.is_initialized():
+                return False
+            return True
+
+        if not is_dist_avail_and_initialized():
+            return
+
+        eval_gather_dicts: List[Dict] = [defaultdict(list)] * dist.get_world_size()
+        dist.all_gather_object(eval_gather_dicts, self.eval_imgs)
+
+        category_gather_list: List[Set] = [set()] * dist.get_world_size()
+        dist.all_gather_object(category_gather_list, self.category_ids)
+
+        keys = set()
+        for eval_imgs in eval_gather_dicts:
+            for key in eval_imgs:
+                keys.add(key)
+        combined_eval_imgs: defaultdict = defaultdict(list)
+        for key in keys:
+            for eval_imgs in eval_gather_dicts:
+                combined_eval_imgs[key] += eval_imgs[key]
+        self.eval_imgs = combined_eval_imgs
+
+        all_category_ids = set()
+        for category_set in category_gather_list:
+            for category_id in category_set:
+                all_category_ids.add(category_id)
+
+        self.category_ids = all_category_ids
+
+    @sync_all_reduce()
+    def compute(self) -> Dict:
+        self._gather_all()
+        self._accumulate()
+
+        results = {
+            "all@0.5": self._summarize(iou_thr=0.5),
```

Comment:
if 0.5 and 0.75 are not values from tresholds, what happen ?

Reply:
It will return -1.